<launch>
    <!-- Arguments -->
    <arg name="robot_namespace" default="/"/>
    <!-- Helper arg to avoid double slashes when robot_namespace is "/" -->
    <arg name="ns_prefix" default="$(eval '' if arg('robot_namespace') == '/' else arg('robot_namespace') + '/')"/>
    <!-- <arg name="cmd_vel_topic" default="/cmd_vel" />
    <arg name="odom_topic" default="odom" /> -->
    <arg name="move_forward_only" default="false"/>
    <arg name="global_frame_id" default="map"/>
    <arg name="laser_topic" default="scan"/>
    
    <arg name="odom_frame_id" value="$(arg ns_prefix)odom"/>

    <arg name="base_frame_id" value="$(arg ns_prefix)base_footprint"/>

    <!-- move_base_flex (MBF) -->
    <node pkg="mbf_costmap_nav" type="mbf_costmap_nav" respawn="false" name="move_base_flex" output="screen" required="true">
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
        <!-- <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" /> -->

        <!-- Costmap 配置 -->
        <rosparam file="$(find scout_navigation)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find scout_navigation)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find scout_navigation)/param/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find scout_navigation)/param/global_costmap_params.yaml" command="load" />
        
        <!-- Move Base Flex 配置（包含 planners 和 controllers） -->
        <rosparam file="$(find scout_navigation)/param/move_base_flex.yaml" command="load" />
        
        <!-- 规划器和控制器参数 -->
        <rosparam file="$(find scout_navigation)/param/dwa_local_planner_params.yaml" command="load" />
        <!-- <rosparam file="$(find scout_navigation)/param/teb_local_planner_params.yaml" command="load" /> -->
        <!-- <rosparam file="$(find scout_navigation)/param/global_planner_params.yaml" command="load" /> -->
        <rosparam file="$(find scout_navigation)/param/navfn_global_planner_params.yaml" command="load" />
        
        <!-- 重要：这些参数必须在 YAML 文件加载之后设置，以覆盖 YAML 中的默认值 -->
        <!-- 设置全局 costmap 参数 -->
        <param name="global_costmap/global_frame" value="$(arg global_frame_id)"/>
        <param name="global_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
        <param name="global_costmap/obstacle_layer/scan/sensor_frame" value="$(arg robot_namespace)/hokuyo_link"/>
        <param name="global_costmap/obstacle_layer/scan/topic" value="$(arg laser_topic)"/>
        <param name="global_costmap/scan/topic" value="$(arg robot_namespace)/$(arg laser_topic)"/>
        <param name="global_costmap/scan/sensor_frame" value="$(arg robot_namespace)/hokuyo_link"/>
        <!-- 设置 robot_obstacles observation source for global costmap -->
        <param name="global_costmap/obstacle_layer/robot_obstacles/topic" value="robot_obstacles_global"/>
        <param name="global_costmap/obstacle_layer/robot_obstacles/sensor_frame" value="base_footprint"/>
        
        <!-- 设置局部 costmap 参数 -->
        <param name="local_costmap/global_frame" value="$(arg odom_frame_id)"/>
        <param name="local_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
        <param name="local_costmap/scan/sensor_frame" value="$(arg robot_namespace)/hokuyo_link"/>
        <param name="local_costmap/obstacle_layer/scan/sensor_frame" value="$(arg robot_namespace)/hokuyo_link"/>
        <param name="local_costmap/obstacle_layer/scan/topic" value="$(arg laser_topic)"/>
        <param name="local_costmap/scan/topic" value="$(arg robot_namespace)/$(arg laser_topic)"/>
        <!-- 设置 robot_obstacles observation source -->
        <param name="local_costmap/obstacle_layer/robot_obstacles/topic" value="robot_obstacles"/>
        <!-- sensor_frame不能以/开头，使用相对frame -->
        <param name="local_costmap/obstacle_layer/robot_obstacles/sensor_frame" value="base_footprint"/>
        
    <!-- move_base_flex 可能需要显式设置 robot_frame 参数 -->
    <param name="robot_frame" value="$(arg base_frame_id)"/>
    
    <!-- Remap topics 到命名空间内的 topics -->
    <!-- cmd_vel: 发布速度命令到命名空间内的 cmd_vel -->
    <remap from="cmd_vel" to="cmd_vel"/>
    <!-- odom: 订阅命名空间内的 odom topic -->
    <remap from="odom" to="odom"/>

    </node>
    
    <!-- 机器人障碍物点云发布器 -->
    <node pkg="scout_navigation" type="robot_obstacle_publisher.py" name="robot_obstacle_publisher" output="screen" ns="$(arg robot_namespace)">
        <param name="robot_namespace" value="$(arg robot_namespace)"/>
        <param name="distance_threshold" value="3.0"/>  <!-- 距离阈值（米） -->
        <param name="robot_radius" value="0.4"/>  <!-- 机器人半径（米） -->
        <param name="publish_rate" value="30.0"/>  <!-- 发布频率（Hz） -->
        <param name="footprint" value="[[-0.32, -0.31], [-0.32, 0.31], [0.32, 0.31], [0.32, -0.31]]"/>  <!-- Scout footprint -->
        <param name="robot_namespaces" value="robot_1,robot_2,robot_3,robot_4,robot_5,robot_6"/>
        <!-- robot_obstacles话题会自动映射到命名空间内 -->
    </node>
</launch>
